File: Chess Javascript Code/Move.js
const {Square} = require("./Square.js");

class Move
{
	//squares pertaining to the main piece of the move
	mainPieceSquareBefore;	//the square the piece started at
	mainPieceSquareAfter;	//the square the piece ended at
	
	string;
	
	constructor(mainPieceSquareBefore, mainPieceSquareAfter)
	{		
		this.mainPieceSquareBefore = mainPieceSquareBefore;
		this.mainPieceSquareAfter = mainPieceSquareAfter;
	}
	
	//a move involving the same squares could have a different string representation across different positions
	//this function generates the move's string for the current position
	generateString(game)
	{
		this.string = game.pieces[this.mainPieceSquareBefore].constructor.makeMoveString(this, game);
	}
	
	toString(game)
	{
		//if the string did not already exist, generate it on the spot
		return this.string ?? ((move)=>{move.generateString(game); return move.string;})(this);
	}
}

class PlainMove extends Move
{
}

class CastleMove extends Move
{
	rookBefore;
	rookAfter;

	constructor(kingBefore, kingAfter, rookBefore, rookAfter)
	{
		super(kingBefore, kingAfter);
		this.rookBefore = rookBefore;
		this.rookAfter = rookAfter;
	}
	
	generateString()
	{
		this.string = (Square.file(this.mainPieceSquareBefore) > Square.file(this.mainPieceSquareAfter))? "O-O-O" : "O-O";
	}
}

class EnPassantMove extends Move
{
	enPassantSquare;
	constructor(pawnBefore, pawnAfter, enPassantSquare)
	{
		super(pawnBefore, pawnAfter);
		this.enPassantSquare = enPassantSquare;
	}
}

class PromotionMove extends Move
{
	promotionPiece;
	constructor(mainPieceSquareBefore, mainPieceSquareAfter, promotionPiece)
	{
		super(mainPieceSquareBefore, mainPieceSquareAfter);
		this.promotionPiece = promotionPiece;
	}
}

module.exports = {
	PlainMove:PlainMove,
	CastleMove:CastleMove,
	EnPassantMove:EnPassantMove,
	PromotionMove:PromotionMove
}

File: Chess C++ Code/Move.cpp
#include <string>
using std::string;

#include "Move.h"
#include "Square.h"
using Square::square_t;

const Move Move::DUMMY_MOVE(Square::DUMMY_SQUARE, Square::DUMMY_SQUARE);

Move::Move(square_t mainPieceSquareBefore, square_t mainPieceSquareAfter) :
	mainPieceSquareBefore(mainPieceSquareBefore), mainPieceSquareAfter(mainPieceSquareAfter)
{}

square_t Move::getMainPieceSquareAfter() const
{
	return this->mainPieceSquareAfter;
}
square_t Move::getMainPieceSquareBefore() const
{
	return this->mainPieceSquareBefore;
}

bool Move::equals(Move& m) const {
	return (this->getMainPieceSquareAfter() == m.getMainPieceSquareAfter()) && (this->getMainPieceSquareBefore() == m.getMainPieceSquareBefore());
}

string Move::toString() {
	return Square::fullString(this->mainPieceSquareBefore) + "->" + Square::fullString(this->mainPieceSquareAfter);
}

const PlainMove PlainMove::DUMMY_PLAINMOVE(Square::DUMMY_SQUARE, Square::DUMMY_SQUARE);

PlainMove::PlainMove(square_t mainPieceSquareBefore, square_t mainPieceSquareAfter) : Move(mainPieceSquareBefore, mainPieceSquareAfter)
{}